routes/bridges/gassistant/index.js

```diff
@@ -0,0 +1,235 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of ThingEngine
+//
+// Copyright 2017-2019 The Board of Trustees of the Leland Stanford Junior University
+//
+// Author: Swee Kiat Lim <sweekiat@cs.stanford.edu>
+//
+// See COPYING for details
+"use strict";
+
+const Q = require('q');
+const express = require('express');
+const passport = require('passport');
+const util = require('util');
+const jwt = require('jsonwebtoken');
+
+const EngineManager = require('../../../almond/enginemanagerclient');
+const { actionssdk, Image, Suggestions, BasicCard, Button, SignIn } = require('actions-on-google');
+// Refer to https://developers.google.com/assistant/conversational/responses for full list of response types
+
+const userUtils = require('../../../util/user');
+const db = require('../../../util/db');
+const model = require('../../../model/user');
+const secret = require('../../../util/secret_key');
+
+var router = express.Router();
+
+class GoogleAssistantDelegate {
+    constructor(locale) {
+        this._buffer = [];
+        this._locale = locale;
+        this._requestSignin = false;
+    }
+
+    send(text, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] === 'string')
+            // If there is already a text reply immediately before, we merge text replies
+            // because Google Assistant limits at most 2 chat bubbles per turn
+            this._buffer[this._buffer.length - 1] += '\n' + text;
+        else
+            this._buffer.push(text);
+    }
+
+    sendPicture(url, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] !== 'string')
+            // If there is no text reply immediately before, we add the URL
+            // because Google Assistant requires a chat bubble to accompany an Image
+            this._buffer.push(url);
+        this._buffer.push(new Image({
+            url: url,
+            alt: url,
+        }));
+    }
+
+    sendRDL(rdl, icon) {
+        this._buffer.push(new BasicCard({
```

Comment:
Do we need fallback for Google Home devices without a screen?

Reply:
Yup. In that case, we will just display the title and text?

Reply:
@gcampax

Reply:
Yeah the question is, do we need to change the code, or does Google handle that?